platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/builder/PlatformBuilder.java

```diff
@@ -383,63 +426,77 @@ public Platform build() {
-                                        ? null
-                                        : addressBookInitializer
-                                                .getPreviousAddressBook()
-                                                .copy());
-            }
+                emergencyRecoveryManager);
 
-            // At this point the initial state must have the current address book set.  If not, something is wrong.
-            if (initialState.get().getState().getPlatformState().getAddressBook() == null) {
-                throw new IllegalStateException("The current address book of the initial state is null.");
-            }
+        final boolean softwareUpgrade = detectSoftwareUpgrade(softwareVersion, initialState.get());
 
-            final SwirldsPlatform platform = new SwirldsPlatform(
-                    platformContext,
-                    keysAndCerts.get(selfId),
-                    recycleBin,
-                    selfId,
-                    appName,
-                    swirldName,
-                    softwareVersion,
-                    initialState.get(),
-                    emergencyRecoveryManager,
-                    preconsensusEventConsumer,
-                    snapshotOverrideConsumer);
-
-            if (firstTimeSetup) {
-                MetricsDocUtils.writeMetricsDocumentToFile(getGlobalMetrics(), getPlatforms(), configuration);
-                getMetricsProvider().start();
-            }
+        // Initialize the address book from the configuration and platform saved state.
+        final AddressBookInitializer addressBookInitializer = new AddressBookInitializer(
+                selfId,
+                softwareVersion,
+                softwareUpgrade,
+                initialState.get(),
+                configAddressBook.copy(),
+                platformContext);
+
+        if (addressBookInitializer.hasAddressBookChanged()) {
+            final State state = initialState.get().getState();
+            // Update the address book with the current address book read from config.txt.
+            // Eventually we will not do this, and only transactions will be capable of
+            // modifying the address book.
+            state.getPlatformState()
+                    .setAddressBook(
+                            addressBookInitializer.getCurrentAddressBook().copy());
+
+            state.getPlatformState()
+                    .setPreviousAddressBook(
+                            addressBookInitializer.getPreviousAddressBook() == null
+                                    ? null
+                                    : addressBookInitializer
+                                            .getPreviousAddressBook()
+                                            .copy());
+        }
+
+        // At this point the initial state must have the current address book set.  If not, something is wrong.
+        if (initialState.get().getState().getPlatformState().getAddressBook() == null) {
```

Comment:
Instead of calling `initialState.get().getState().getPlatformState().getAddressBook()` twice, you could move this check below the definition of `final AddressBook currentAddressBook`, and then do this:

```
if (currentAddressBook == null) {
            throw new IllegalStateException("The current address book of the initial state is null.");
        }
```

Reply:
done